Add `kb-sdk shell` command

jayrbolton opened the pull request:
Here is my first shot at adding some Java. @MrCreosote let me know what feedback you have.

This opens the docker shell as a root user, which I know Shane didn't want. But it works well for me. We can update this PR later with an alternative UID. Also I think we could remove the `run_bash.sh` script once this update looks good.

jayrbolton committed:
Add the command in ModuleBuilder and the class handler in mobue/shell_launcher/ShellLauncher.java

jayrbolton committed:
Add comments to ShellLauncher#exec

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+import us.kbase.mobu.util.TextUtils;
+import us.kbase.mobu.ModuleBuilder;
+
+public class ShellLauncher {
```
Needs javadocs

MrCreosote replied:
blank line after class declaration

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+import us.kbase.mobu.ModuleBuilder;
+
+public class ShellLauncher {
+    File moduleDir;
```
Class fields should be `private final` unless there's a good reason otherwise

MrCreosote replied:
Use `Path` from nio, File is out of date.

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+public class ShellLauncher {
+    File moduleDir;
+    String moduleName;
+    String workDir;
```
Use `Path`

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+
+    public Process exec() throws IOException, InterruptedException {
+        // This command looks like:
+        // `docker run -i -t -v WORKDIR:/kb/module/work test/module:latest bash`
```
Are there long forms of the flags? Would be easier to read if so. If not, document what the flags mean

MrCreosote commented on src/java/us/kbase/mobu/ModuleBuilder.java:
```diff
@@ -29,6 +29,7 @@
 import us.kbase.mobu.util.ProcessHelper;
```
Is there any way to automatically test this?

MrCreosote replied:
Still have the same question here.

MrCreosote commented on src/java/us/kbase/mobu/ModuleBuilder.java:
```diff
@@ -96,6 +98,10 @@ public static void main(String[] args) throws Exception {
         jc.addCommand(RUN_COMMAND, runArgs);
 
+        // add the 'shell' command
+        ShellCommandArgs shellArgs = new ShellCommandArgs();
```
`final`

MrCreosote commented on src/java/us/kbase/mobu/ModuleBuilder.java:
```diff
@@ -437,6 +445,18 @@ private static int runRunCommand(RunCommandArgs runArgs, JCommander jc) {
         }
     }
 
+    private static int runShellCommand(ShellCommandArgs shellArgs, JCommander jc) {
```
`final` args

MrCreosote commented on src/java/us/kbase/mobu/ModuleBuilder.java:
```diff
@@ -437,6 +445,18 @@ private static int runRunCommand(RunCommandArgs runArgs, JCommander jc) {
+            launcher.exec();
+            return 0;
+        } catch (Exception e) {
+            e.printStackTrace();
```
Add a verbose argument like the other commands, and only print the trace if true

MrCreosote replied:
Alternatively just pass the exception and verbose boolean to showError() and put all the error display logic in one spot

MrCreosote commented on src/java/us/kbase/mobu/ModuleBuilder.java:
```diff
@@ -137,7 +143,9 @@ public static void main(String[] args) throws Exception {
 	    } else if (jc.getParsedCommand().equals(RUN_COMMAND)) {
 	        returnCode = runRunCommand(runArgs, jc);
-	    }
+	    } else if (jc.getParsedCommand().equals(SHELL_COMMAND)) {
```
There's some inconsistent indenting going on here

jayrbolton replied:
Unfortunately there is a big mix of tabs and spaces throughout ModuleBuilder.java. I changed it to match the "formatting" of the rest of this section (ie. a tab character and then spaces afterward).

eg. line 151 is `<tab-character><four-spaces><tab-character>System.exit(returnCode);`

:man_shrugging:

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+        moduleDir = DirUtils.findModuleDir();
+        String kbaseYml = TextUtils.readFileText(new File(moduleDir, "kbase.yml"));
+        @SuppressWarnings("unchecked")
+        Map<String,Object> config = (Map<String, Object>)new Yaml().load(kbaseYml);
```
Do `new Yaml(new SafeConstructor())`

MrCreosote replied:
If the yaml is corrupted or isn't a map this will throw a nasty error

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+        String kbaseYml = TextUtils.readFileText(new File(moduleDir, "kbase.yml"));
+        @SuppressWarnings("unchecked")
+        Map<String,Object> config = (Map<String, Object>)new Yaml().load(kbaseYml);
+        moduleName = (String)config.get("module-name");
```
Check for null in case the yaml file is corrupted

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+    String workDir;
+
+    public ShellLauncher() throws IOException {
+        moduleDir = DirUtils.findModuleDir();
```
It seems like all this stuff is something that's going to happen over and over - getting the module dir, module name, and work dir. Maybe make a class that does all that stuff in one place?

MrCreosote commented on src/java/us/kbase/mobu/shell_launcher/ShellLauncher.java:
```diff
@@ -0,0 +1,48 @@
+        workDir = Paths.get(moduleDir.toPath().toString(), "test_local", "workdir").toString();
+    }
+
+    public Process exec() throws IOException, InterruptedException {
```
Returning the process seems really weird to me. I would make the method `void` unless there's a good reason otherwise.

MrCreosote commented on src/java/us/kbase/mobu/ModuleBuilder.java:
```diff
@@ -29,6 +29,7 @@
 import us.kbase.mobu.util.ProcessHelper;
```
Have you tested with mac & linux, and do you see any of the issues described in the `--user` ticket?

jayrbolton replied:
Running as root causes any files you create in the container to be owned by "root" in the host. To me this is not a problem and is something I would expect with docker. But Shane said it was a usability issue for people. I'll to talk to Shane about it soon in person.

MrCreosote replied:
What happens if you run the container as root and produce files owned by root on a system where you don't have root access? Seems like you wind up with a bunch of files in your home directory you can't touch.

jayrbolton commented:
Thanks for the review. I took a lot of the code around getting the module name/directory from elsewhere ([for example here](https://github.com/kbase/kb_sdk/blob/master/src/java/us/kbase/mobu/renamer/ModuleRenamer.java#L45)). It felt pretty overly complicated and seems like it should be a lot easier. @MrCreosote maybe you have better boilerplate for passing/fetching module name/directory from the `ShellLauncher` class?

MrCreosote commented:
Not off the top of my head. The first thing I'd do is isolate all that stuff in a `ModuleInfo` class (or something) so if there's a way to simplify it can be done in one place rather than all over.

jayrbolton committed:
Add some javadocs; don't return the Process from exec(); make instance variabled

jayrbolton committed:
Add in ModuleInfo for fetching module name/directories using nio

jayrbolton committed:
Add slightly better error handling/messaging for loading the yaml config file

jayrbolton committed:
Add an explicit check on the kbase.yml "module-name" key

jayrbolton committed:
Add some "final" flags and edit some comments

jayrbolton committed:
Add tab character to try to be consistent with the rest of the conditional section

jayrbolton committed:
Make more variables final; add a verbose flag for the shell command

jayrbolton committed:
Replace all tab characters with four spaces inside ModuleBuilder.java

jayrbolton committed:
Clean up error message

jayrbolton commented:
@MrCreosote I made changes and would be ready for a re-review at any point, no hurry. Change notes:
- Added javadocs comments
- Returning void rather than Process in ShellLauncher#exec()
- Converted a lot of my variables to final (I think all of them)
- Added a `ModuleInfo` class that gets module config and directories. Has some better error handling.
- Replaced all tab characters in ModuleBuilder with four spaces

Other notes:
- I will test on a mac soon
- Can do a RELASE_NOTES.txt update for this and increase patch semver. Not sure what conventions you all have here. 
- Lots of other files mix tabs and spaces. I can fix all of them in another PR.
- There is a lot of opportunity to refactor code all over the place here. Will be tricky to balance big refactors vs incremental changes on this repo

MrCreosote commented:
> I will test on a mac soon
We should do that before merge.

>Can do a RELASE_NOTES.txt update for this and increase patch semver. Not sure what conventions you all have here.

Just follow the conventions in the current file, but make the release date TBD. We'll add that when we merge to master.

> Lots of other files mix tabs and spaces. I can fix all of them in another PR.

Dooo eet

> There is a lot of opportunity to refactor code all over the place here. Will be tricky to balance big refactors vs incremental changes on this repo

Refactor as you go IMO, otherwise you'll be refactoring for the next few months. If you're working on a chunk of code, make it better.

jayrbolton committed:
Add RELEASE_NOTES.txt updates for the added features

jayrbolton commented:
Still todo on this one:
- [ ] test on mac
- [ ] take some kind of stab at making tests for the classes created here (ModuleInfo, at least)

jayrbolton committed:
Merge branch 'develop' into add-shell-command

jayrbolton commented:
@sychan This could potentially be merged. I've tested it quite a bit, including on mac. The remaining thing I wanted to do was take a stab at writing tests for the `ModuleInfo` class I added here, but didn't make a ton of headway. Trying to write tests for that class might not be worth it right now

jayrbolton committed:
Merge branch 'develop' into add-shell-command

jayrbolton commented:
Diff without the whitespace changes to ModuleBuilder: https://github.com/kbase/kb_sdk/pull/314/files?w=1

jayrbolton commented:
@scanon maybe you have feedback on this one

jayrbolton commented:
:man_shrugging:
